### PR TITLE
[#2801] Create new python distribution.

### PR DESCRIPTION
**Pull request by adiroiban**
# Scope

This is a new way of creating and distributing the python package.

To not break the current build system a new repo was created. 

The python distribution will only contain python + pip + setuptools. This should be enough to install other packages like pyopenss/pysqlite... etc

pyopenssl/pysqlite will be built from separate repos, each module will have its own repo and own version.
# Changes

Buildmaster was updated to include a builder for the new python-distribution http://build.chevah.com/waterfall?tag=python-distribution  ...for now it has Ubuntu and Win XP.
We might want to replace Ubuntu with Linux generic ... but I want to see how the linux generic is handled in python-package and brink.

The new paver.sh should just take care of downloading the python distribution and then invoke a few hook based on paver.conf

The new build system re-user part of the old build system just that it no longer keep all source code in the repo and does not build pysqlite/pyopenssl/pycryto...etc

I don't know what is the best way to patch the upstream version... for now, chevahbs has a separate 'patch' step which can be used to execute command which modified the upstream version.

It no longer has production/testing as each build has a specific version.

It requires a modified paver.sh and will also require a modified brink

DEFAULT_VALUE file is still created as I don't know how to call the script back from python.

montage.sh detect_os to show those values to stdout without writing the file

paver.sh and paver.conf were renamed to montage.sh and montage.conf as they no longer depend on paver.

montage.sh can by used with any other build system... including python setup.py.
All brink specific code and chevah custom build system code is included in the montage.conf file. 

The python-distro is now the source for montage.sh

The new pip version automatically caches the files so we no longer need to cache... the cache is still used for python binaries... but in the future we might want to move this to per user cache, rather than per repo

A new brink version was created to support the new build system https://github.com/chevah/brink/pull/70
# How to test

reviewers: @dumol 

check that the new build system make sense and is well documented.

`./venv.sh`
`./venv.sh virtualenv PATH-TO-NEW-VENV`
`./venv.sh clean`

First remote build was successful http://build.chevah.com/waterfall?tag=python-distribution

There are still many things to be fixed but I have created separate tickets for them https://github.com/chevah/python-distribution/issues

Thanks!

### CODE REVIEW

**Comment by adiroiban** [comment]
I think that this is ready for review... we might need to discuss this over voice or irc as the changes are big.

thanks!

needs-review

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
+python-distribution
+===================
+
+This code is here to generate a portable CPython distribution on all the
+platforms supported by the Chevah project... and not only.
+
+It is designed to by the Chevah Buildbot script to automatically generate and
+publish the binary python distribution.
```
Comment (dumol) [inline README.rst]: Can't make sense of the previous sentence.

File: chevah_build [hunk 2]
```diff
@@ -0,0 +1,297 @@
+#!/usr/bin/env bash
+#
+# Chevah Build Script for CPython distribution.
+#
+
+PYTHON_UPSTREAM_VERSION='2.7.8'
+PYTHON_BUILD_VERSION='.c1'
```
Comment (dumol) [inline chevah_build]: For our minor versions I would suggest using another delimiter, such as `-`.

File: chevah_build [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+UBUNTU_PACKAGES="gcc make libssl-dev zlib1g-dev m4 texinfo"
+RHEL_PACKAGES="gcc make openssl-devel zlib-devel m4 texinfo"
+SLES_PACKAGES="gcc make libopenssl-devel zlib-devel m4 texinfo"
+# List of OS packages requested to be installed by this script.
+INSTALLED_PACKAGES=''
+# For the moment, we don't install anything on OS X, Solaris, AIX and
+# unsupported Linux distros. The build requires a C compiler, GNU make, m4,
+# makeinfo (from texinfo, optional) and the header files for OpenSSL and zlib.
+
+PYTHON_VERSION=$PYTHON_UPSTREAM_VERSION$PYTHON_BUILD_VERSION
```
Comment (dumol) [inline chevah_build]: Perhaps we should use the delimiter here, for clarity?

File: chevah_build [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+PYTHON_BUILD_VERSION='.c1'
+LIBFFI_UPSTREAM_VERSION='3.2.1'
+
+# Import shared code.
+. ./functions.sh
+
+# List of OS packages required for building Python.
+UBUNTU_PACKAGES="gcc make libssl-dev zlib1g-dev m4 texinfo"
+RHEL_PACKAGES="gcc make openssl-devel zlib-devel m4 texinfo"
+SLES_PACKAGES="gcc make libopenssl-devel zlib-devel m4 texinfo"
```
Comment (dumol) [inline chevah_build]: The list of packages needs to be updated from the master of `python-package`.

**Comment by dumol** [comment]
As discussed on IRC today…

**needs-changes**

File: chevah_build (hunk 2 above)
Comment (adiroiban) [inline chevah_build]: true. I went for + and documented it in the README

File: README.rst (hunk 1 above)
Comment (adiroiban) [inline README.rst]: true. I tried to reprhase it. please check the new version.

File: chevah_build (hunk 3 above)
Comment (adiroiban) [inline chevah_build]: We have boht + and - delimiter... so for now we are stuck with this.

The - delimiter is used to signal testing version and we can skip these version when doing the mirror and also clean them after a while.... Versions with + delimiter should be kept almost forever

File: chevah_build (hunk 4 above)
Comment (adiroiban) [inline chevah_build]: we  will no longer build readline/sqlite/pysqlite from this repo so we don't need to install the ncurses lib here.

**Comment by adiroiban** [comment]
For now, I left the function.sh as a separate file as it is used in chevahbs.sh ... but we can look at remove it into a separate ticket.

The scripts were renamed to have the .sh extensions.

montage.sh was renamed to venv.sh and I tried to document it. For now, I would like to try to keep a single file for montage.sh, which will include its documentation. montage.conf is optional....
if having code in montage.conf is a big problem we can later split in into montage.conf and montage.hooks. If someone wants to have the hooks in a separate file he/she can already source montage.hooks from a custom montage.conf

The default python distribution will be built without readline/libedit. We will have a separate repo which will build this binary module.

I went for 2.7.8+chevah1 as production version. ... later we can improve the build script to prevent overwriting public versions and only allow overwriting of pre-release versions

Please check latest changes. I have generated binaries for generic linux so you can use them to see if brink/paver works.

Many thanks for the review and the feedback!

needs-change

**Comment by adiroiban** [comment]
needs-review

**Comment by dumol** [comment]
I've tried `./chevah_build.sh build` and `./chevah_build.sh build` on my Hardened Gentoo system and things worked. I think `README.rst` should be updated accordingly as, for now, it mentions the script by its old name.

Not sure about the stuff mentioned in the _How to test_ section… I've tried using `venv.sh` instead of `montage.sh`, but invoking it by itself didn't work for me:

```
% ./venv.sh
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/dumol/work/chevah/python-distribution/build-runtime/lib/python2.7/site-packages/paver/tasks.py", line 883, in main
    _launch_pavement(args)
  File "/home/dumol/work/chevah/python-distribution/build-runtime/lib/python2.7/site-packages/paver/tasks.py", line 851, in _launch_pavement
    exec(compile(source, environment.pavement_file, 'exec'), mod.__dict__)
  File "pavement.py", line 20, in <module>
    from brink.pavement_commons import (
ImportError: No module named brink.pavement_commons
```

While this did end successfully: `./venv.sh virtualenv ../tmp`

This one worked also: `./venv.sh clean`. But it didn't clear the cache… Should it?

**needs-changes**

**Comment by adiroiban** [comment]
`ImportError: No module named brink.pavement_commons` 

it looks like you might already have an invalid python distro.

Please try `venv.sh clean` and `venv.sh` again.

Are you on the latest version?

I have update the readme.

clean should not delete the cache folder. Python is now versioned so we no longer need to clean it....
we might want to add a clean-dist command which will also delete the cache.

thanks for the review

needs-changes

**Comment by dumol** [comment]
I guess you wanted this again:

**needs-review**

**Comment by dumol** [comment]
After `venv.sh clean` things are fine for me too. Thanks!

**changes-approved**

**Pull request closed without merging**